src/units.rs

```diff
@@ -122,231 +115,128 @@ impl<H: Hasher> PreUnit<H> {
     pub(crate) fn n_members(&self) -> NodeCount {
         self.control_hash.n_members()
     }
-}
 
-impl<H: Hasher> From<PreUnit<H>> for NotificationOut<H> {
-    fn from(pu: PreUnit<H>) -> Self {
-        NotificationOut::CreatedPreUnit(pu)
+    pub(crate) fn creator(&self) -> NodeIndex {
+        self.coord.creator()
+    }
+    pub(crate) fn round(&self) -> Round {
+        self.coord.round()
+    }
+    pub(crate) fn control_hash(&self) -> &ControlHash<H> {
+        &self.control_hash
     }
 }
 
-#[derive(Clone, Debug, Default, PartialEq, Encode, Decode)]
-pub(crate) struct Unit<H: Hasher> {
-    pub(crate) creator: NodeIndex,
-    round: UnitRound,
-    pub(crate) hash: H::Hash,
-    pub(crate) control_hash: ControlHash<H>,
+///
+#[derive(Debug, Default, Clone, Encode, Decode)]
+pub(crate) struct FullUnit<H: Hasher, D: Data> {
+    pre_unit: PreUnit<H>,
```

Comment:
Why not store Unit and skip encoding hash there?

Reply:
Then we would have to initialize the hash of that Unit after constructing the FullUnit, which seems like a very bad practice.
Also, for me the semantics of codec(skip) is that the field is reduntant and can be inferred from other fields. And in Unit, this does not hold for the field hash. My solution is the lmost logical I can see.